---
title: "Using the mePipe package"
author: "Peter Humburg"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette: 
    toc: TRUE
vignette: >
  %\VignetteIndexEntry{Using the mePipe package}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

# Introduction
The *mePipe* package provides a command-line interface to commonly used `MatrixEQTL` 
functionality and augments this with additional features that are useful for a typical 
eQTL analysis. These include the use (and automatic selection) of principal components 
as covariates to account for confounding variation, accounting for LD structure and 
running on a SGE cluster rather than a single processor.

The most commonly used options are described below. A full list of available options 
for the installed version is available from the command-line via `runMatrixEQTL.R --help`.

In the following `expression_file` refers to the name (including the relevant path 
to the file on the file system if it isn't located in the working directory) of a text 
file containing the gene expression data. Similarly, `genotype_file` refers to the 
file with the genotyping information.

# File formats
## Input files
Input files (gene expression measurements, genotypes and covariates) are plain text 
files with samples in columns and variables (genes, SNPs or other variables) in rows. 
*Note that these data are often stored with samples in rows. 
See TransposingLargeFiles on how to convert from one format to the other.*

By default files are assumed to be `tab` delimited but any delimiter can be used as 
long as it is consistent between all input files. Use the `--delim` option to set 
the character used as delimiter. For example,

```sh
runMatrixEQTL.R --delim=, expression_file genotype_file
```
will assume that the input files are comma separated, i.e. csv, files.

Input files may have any number of row and column headers as long as these are 
consistent across files. The default is to assume a single row of column names and a 
single column of row names. Use `--rowskip` and `--colskip` to change this. 
The following example would work for files with two rows of column (sample) labels 
and three columns of row (variable) labels.

```sh
runMatrixEQTL.R --rowskip=2 --colskip=3 expression_file genotype_file
``` 

## Genotype format
Genotypes have to be encoded numerically with a single entry for each site per sample. 
For this `0`, `1` and `2` should be used in the usual fashion. This encoding is mandatory 
if LD calculations are carried out (see [below](#ldblocks)). For the additive model 
(without LD computations) any numerical value is acceptable and this can be used to account 
for genotyping (or imputation) uncertainty. 

### Converting from other formats
Often genotyping information is stored in a different format to the one required by 
Matrix-eQTL. Some scripts to help with the conversion from common formats are available:
[[ListTagged(fileConversion mePipe)]]

## Output files {#mainout}
The main output [produced by Matrix-eQTL](http://www.bios.unc.edu/research/genomic_software/Matrix_eQTL/manual.html#off)
are tab-delimited text files with SNP-gene associations. These files have five columns 
listing SNP ID, gene (or probe) name, value of the test statistic, p-value and FDR. 
The p-value column is the one referred to by the threshold that is set with the `--pthreshold`
 option. Note that only SNP-gene pairs with a p-value below `pthreshold` are listed in the 
 output. All other combinations are discarded by Matrix-eQTL.


# Basic analysis
A basic analysis that assumes an additive effect of genotype on gene expression and 
doesn't distinguish between cis- and trans-effects can be run as follows.

```sh
runMatrixEQTL.R --output=foo  expression_file genotype_file
```
This will run a simple additive model without covariates and report all associations 
with an (uncorrected) p-value of less than \(10^{-5}\) in a file called `foo`. 

The p-value threshold can be changed with the `--pthreshold` option. 
The command below will set the threshold to \(10^{-8}\).

```sh
runMatrixEQTL.R --output=foo --pthreshold=1e-8 expression_file genotype_file
```
Relaxing the p-value threshold will increase the number of reported associations 
(and run-time!).

To test for dominant as well as additive effects use the `--anova` flag:

```sh
runMatrixEQTL.R --output=foo --anova expression_file genotype_file
```

## Output
All the examples in this section produce a single text file `foo` in the working directory 
with [standard Matrix-eQTL output](#mainout).

In addition The R object generated by `Matrix-eQTL` is saved in a file called `foo.rdata`. 
This may be useful for some further processing or plotting of the data but is usually not 
needed.


# Position aware analysis
To distinguish between cis- and trans-associations two additional input files are 
required to provide genomic locations of SNPs (this file will be referred to as `snp_pos` 
in the following) and genes (`gene_pos`). Both files should be either tab or space delimited 
and contain a single header line. The SNP position file should have three columns 
corresponding to SNP ID, chromosome and location respectively. The gene position file 
should have four columns with gene name, chromosome and gene start and end coordinates.

A position aware analysis with default parameters can then be run like this:

```sh
runMatrixEQTL.R --output=foo  --snpspos=snp_pos --genepos=gene_pos expression_file genotype_file
```
This considers all SNPs that are located less than 1MB away from a given gene to be in 
cis to this gene. All other SNPs are assumed to be in trans. The size of the cis window 
can be adjusted with the `--cisDist` option. Matrix-eQTL uses separate p-value thresholds 
for cis- and trans-associations. The cis threshold can be set with the `--cisthreshold` 
option. The `--pthreshold` option is used for trans associations.

The following would run the same analysis as before but allowing for up to 5MB between 
SNPs and genes for cis effects and using a p-value threshold of \(10^{-4}\) 
(instead of the default of \(10^{-3}\)).

```sh
runMatrixEQTL.R --output=foo  --snpspos=snp_pos --genepos=gene_pos --cisDist=5e6 --cisthreshold=1e-4 expression_file genotype_file
```


## Output
The position aware analysis produces two output files in the usual [Matrix-eQTL output format](#mainout). 
One for trans associations (called `foo` in the examples above) and one for the 
cis associations (`foo.cis`).

# Using covariates
It is often desirable to include additional covariates, i.e. variables other than the SNP
to be tested, into the analysis. Suitably chosen covariates can explain some of the 
(often substantial) variation in gene expression observed between individuals that is not due 
to genetic effects. Several mechanisms are provided by *mePipe* to either [utilise existing 
files](#predefined-covariates) with data for covariates or to [generate a file](#pca-covariates) 
with covariates from the data.

## Predefined covariates 
Often additional data are available for the samples in a study. For example,
gender and age are commonly recorded and some studies come with a substantially
larger set of phenotypic information. Some of these variables may have a large
impact on gene expression. Including relevant variables in the analysis allows
the corresponding variation in gene expression to be attributed properly, 
preventing it from obscuring a, possibly much weaker, SNP effect.

The `--covariate` option can be used to provide a file with one variable per line,
using the same separator as the other input files. All variables contained in
this file will be included in the model for each gene/SNP pair.

## Selecting principle components as covariates {#pca-covariates}
Although it is useful to account for known covariates, the sources of variation
in gene expression are often unknown. One strategy to address this issue relies
on the use of principle component analysis (PCA) to identify the main directions of 
variation. Including the predominant principle components (PC) as covariates
in the analysis has the advantage that no detailed knowledge of possible
confounders is required. It does require, however, a decision on how many PC should
be included. With *mePipe* it is possible to carry out a PCA and automatically
select a suitable number of PC to include as covariates. The selection process
is based on a heuristic that seeks to maximise the number of significant eQTL.
Optionally PC that exhibit significant correlation with genotypes may be excluded
from the selection process.

A number of command-line options are available to control the use of PC covariates.
The `--pcacov` option provides the name of a file with all principle components. This
defaults to the ouput file prefix followed by `_pca_covariates.txt` and will be 
generated automatically, using `prcomp`, if it doesn't exist. If the number of 
PC that should be included is known it can be provided directly using the `--ciscov`,
`--transcov` and `--allcov` arguments for cis, trans and combined analysis respectively. 

If the number of PC to use is to be determined from the data the `--selectcov` flag
has to be included in the command-line. The `--mincov`, `--maxcov`
and `--stepcov` arguments can then be used to specify the range of covariates that should be 
tested for their performance. This will determine the number of significant associations
for every `stepcov` PC between `mincov` and `maxcov`. For example, if these variables are set 
to 5, 5 and 20 respectively, models including 5, 10, 15 and 20 PC will be evaluated. 
The output for all evaluated models is stored in the directory indicated by `--covout`
and output for the model chosen as the best will be copied to the files indicated by
`--output`. This process also produces a plot, located in the file *CovVsEQTL.pdf*
in the `covout` directory, of the number of identified eQTL for each model. The 
`--covthreshold` argument can be used to define an FDR significance threshold for associations
to be included in this plot. It is recommend to check this plot to confirm that 
the chosen number of covariates is sensible. If a different choice seems more 
appropriate the corresponding results can simply be obtained from the files in `covout`.

When using principle components as covariates it is assumed that the major sources of
variation observed in gene expression between individuals are not genetic. Although
this appears to be the case in general, there are instances where some PC show
significant correlation with genotypes. It may be desirable not to include such
PC as covariates in the model. Similarly, it makes sense to exclude any PC that 
is correlated with predefined covariates (provided via `--covariates`). Both of 
these can be accomplished by including the `--filtercov` flag in the command-line.
The relevant FDR threshold can be set with the `--excludecov` argument. This will 
generate an additional file containing only the remaining principle components.  

# Accounting for LD structure

## Resolving multiple cis-associations for the same gene
Once a cis-eQTL has been identified the question may arise whether the observed signal
is due to a single SNP or whether there are indeed several SNPs that contribute
independently to the modulation of gene expression. Due to the LD around the observed 
lead SNP and possibility that independent signals may be masking each other, this
is not a trivial question. One commonly used way to further investigate this is
through conditional analysis. This involves fitting a model that contains the lead SNP
as well as one other candidate SNP. This then provides an estimate of the effect of 
the additional SNP while controling for the effect of the lead SNP. If a sufficiently 
strong signal exists that isn't explained by the lead SNP, this process will result
in a number of significant hits. From these the one with the smallest p-value
can agin be chosen as (secondary) lead SNP. This process can then be iterated (now
using a model including all previously identified lead SNPs) until no significant 
signal remains. With *mePipe* this analysis can be carried out by using the `--multiPeak`
flag. The p-value threshold up to which remaining associations should be considered 
significant can be set with the `--multiPvalue` argument.

Sometimes it is convenient to just run this analysis for a specific gene that is of 
particular interest. This can be enabled with the `--multiGene` argument. 
This is particularly useful when combined with the `--ldOnly` flag that allows 
the use of an existing primary analysis, only running additional LD related computations
(like the multi peak analysis).

It is important to note a few limitations of this approach. Firstly, this approach
can produce a substantial number of false positives if the data don't contain
a good tag SNP. In this situation the observed lead SNP will only explain a fraction of
the underlying signal and the conditional analysis may highlight a large number of 
additional SNPs, each explaining a small part of the same underlying signal. It is
therefore recommended to apply this approach to data that include imputed genotypes.
Secondly, including two SNPs in strong LD into the same model is unlikely to give
accurate results due to colinearity of the two variables. Of course such SNPs are also 
unlikely to provide two independent sources of variation. It therefore is prudent to 
exclude SNPs that are strongly correlated with previously identified lead SNPs. This
can significantly reduce the time required for these computations, especially when 
working with imputed data. The corresponding \(R^2\) threshold can be set with the 
`--ldR2` option.  

## Identifying LD blocks {#ldblocks}

# Running mePipe on SGE {#oncluster}
## Setting up the environment
All dependencies have to be available on the compute nodes. Typically, using an R 
library that is shared between all nodes is sufficient. A notable exception is the 
[XML](http://cran.r-project.org/web/packages/XML/index.html) package, as it relies on 
XML libraries to be available on the system. An R session running on a compute node
has to be able to find a version of the XML library that matches the one the 
package was compiled against.

## Submitting jobs
To run *mePipe* on the cluster simply add the `--cluster` to the command line when invoking 
`runMatrixEQTL.R`. For example, to run the above position aware analysis on the cluster 
the command line would be

```sh
runMatrixEQTL.R --output=foo  --snpspos=snp_pos --genepos=gene_pos --cisDist=5e6 --cisthreshold=1e-4 --cluster expression_file genotype_file
```
This will automatically submit jobs to the compute nodes and collate results at the end. 
Note that the script will continue to run until all jobs have completed. It may therefore 
be advisable to use [nohup](http://en.wikipedia.org/wiki/Nohup).


# Plotting the p-value distribution
