---
title: "Using the mePipe package"
author: "Peter Humburg"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette: 
    toc: TRUE
vignette: >
  %\VignetteIndexEntry{Using the mePipe package}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

# Introduction
The *mePipe* package provides a command-line interface to commonly used `MatrixEQTL` 
functionality and augments this with additional features that are useful for a typical 
eQTL analysis. These include the use (and automatic selection) of principal components 
as covariates to account for confounding variation, accounting for LD structure and 
running on a SGE cluster rather than a single processor.

The most commonly used options are described below. A full list of available options 
for the installed version is available from the command-line via `runMatrixEQTL.R --help`.

In the following `expression_file` refers to the name (including the relevant path 
to the file on the file system if it isn't located in the working directory) of a text 
file containing the gene expression data. Similarly, `genotype_file` refers to the 
file with the genotyping information.

# File formats
## Input files
Input files (gene expression measurements, genotypes and covariates) are plain text 
files with samples in columns and variables (genes, SNPs or other variables) in rows. 
*Note that these data are often stored with samples in rows. 
See TransposingLargeFiles on how to convert from one format to the other.*

By default files are assumed to be `tab` delimited but any delimiter can be used as 
long as it is consistent between all input files. Use the `--delim` option to set 
the character used as delimiter. For example,

```sh
runMatrixEQTL.R --delim=, expression_file genotype_file
```
will assume that the input files are comma separated, i.e. csv, files.

Input files may have any number of row and column headers as long as these are 
consistent across files. The default is to assume a single row of column names and a 
single column of row names. Use `--rowskip` and `--colskip` to change this. 
The following example would work for files with two rows of column (sample) labels 
and three columns of row (variable) labels.

```sh
runMatrixEQTL.R --rowskip=2 --colskip=3 expression_file genotype_file
``` 

## Genotype format
Genotypes have to be encoded numerically with a single entry for each site per sample. 
For this `0`, `1` and `2` should be used in the usual fashion. This encoding is mandatory 
if LD calculations are carried out (see [below](#ldblocks)). For the additive model 
(without LD computations) any numerical value is acceptable and this can be used to account 
for genotyping (or imputation) uncertainty. 

### Converting from other formats
Often genotyping information is stored in a different format to the one required by 
Matrix-eQTL. Some scripts to help with the conversion from common formats are available:
[[ListTagged(fileConversion mePipe)]]

## Output files {#mainout}
The main output [produced by Matrix-eQTL](http://www.bios.unc.edu/research/genomic_software/Matrix_eQTL/manual.html#off)
are tab-delimited text files with SNP-gene associations. These files have five columns 
listing SNP ID, gene (or probe) name, value of the test statistic, p-value and FDR. 
The p-value column is the one referred to by the threshold that is set with the `--pthreshold`
 option. Note that only SNP-gene pairs with a p-value below `pthreshold` are listed in the 
 output. All other combinations are discarded by Matrix-eQTL.


# Basic analysis
A basic analysis that assumes an additive effect of genotype on gene expression and 
doesn't distinguish between cis- and trans-effects can be run as follows.

```sh
runMatrixEQTL.R --output=foo  expression_file genotype_file
```
This will run a simple additive model without covariates and report all associations 
with an (uncorrected) p-value of less than \(10^{-5}\) in a file called `foo`. 

The p-value threshold can be changed with the `--pthreshold` option. 
The command below will set the threshold to \(10^{-8}\).

```sh
runMatrixEQTL.R --output=foo --pthreshold=1e-8 expression_file genotype_file
```
Relaxing the p-value threshold will increase the number of reported associations 
(and run-time!).

To test for dominant as well as additive effects use the `--anova` flag:

```sh
runMatrixEQTL.R --output=foo --anova expression_file genotype_file
```

## Output
All the examples in this section produce a single text file `foo` in the working directory 
with [standard Matrix-eQTL output](#mainout).

In addition The R object generated by `Matrix-eQTL` is saved in a file called `foo.rdata`. 
This may be useful for some further processing or plotting of the data but is usually not 
needed.


# Position aware analysis
To distinguish between cis- and trans-associations two additional input files are 
required to provide genomic locations of SNPs (this file will be referred to as `snp_pos` 
in the following) and genes (`gene_pos`). Both files should be either tab or space delimited 
and contain a single header line. The SNP position file should have three columns 
corresponding to SNP ID, chromosome and location respectively. The gene position file 
should have four columns with gene name, chromosome and gene start and end coordinates.

A position aware analysis with default parameters can then be run like this:

```sh
runMatrixEQTL.R --output=foo  --snpspos=snp_pos --genepos=gene_pos expression_file genotype_file
```
This considers all SNPs that are located less than 1MB away from a given gene to be in 
cis to this gene. All other SNPs are assumed to be in trans. The size of the cis window 
can be adjusted with the `--cisDist` option. Matrix-eQTL uses separate p-value thresholds 
for cis- and trans-associations. The cis threshold can be set with the `--cisthreshold` 
option. The `--pthreshold` option is used for trans associations.

The following would run the same analysis as before but allowing for up to 5MB between 
SNPs and genes for cis effects and using a p-value threshold of \(10^{-4}\) 
(instead of the default of \(10^{-3}\)).

```sh
runMatrixEQTL.R --output=foo  --snpspos=snp_pos --genepos=gene_pos --cisDist=5e6 --cisthreshold=1e-4 expression_file genotype_file
```


## Output
The position aware analysis produces two output files in the usual [Matrix-eQTL output format](#mainout). 
One for trans associations (called `foo` in the examples above) and one for the 
cis associations (`foo.cis`).

# Using covariates
It is often desirable to include additional covariates, i.e. variables other than the SNP
to be tested, into the analysis. Suitably chosen covariates can explain some of the 
(often substantial) variation in gene expression observed between individuals that is not due 
to genetic effects. Several mechanisms are provided by *mePipe* to either [utilise existing 
files](#predefined-covariates) with data for covariates or to [generate a file](#pca-covariates) 
with covariates from the data.

## Predefined covariates {#pca-covariates}

## Selecting principle components as covariates

## Testing for interactions between SNPs and a single covariate

## Output

# Accounting for LD structure {#ldblocks}

## Output

# Running mePipe on SGE {#oncluster}
## Setting up the environment
The [XML](http://cran.r-project.org/web/packages/XML/index.html) package relies on 
XML libraries to be available on the system. These have to be available on the compute 
nodes.

## Submitting jobs
To run *mePipe* on the cluster simply add the `--cluster` to the command line when invoking 
`runMatrixEQTL.R`. For example, to run the above position aware analysis on the cluster 
the command line would be

```sh
runMatrixEQTL.R --output=foo  --snpspos=snp_pos --genepos=gene_pos --cisDist=5e6 --cisthreshold=1e-4 --cluster expression_file genotype_file
```
This will automatically submit jobs to the compute nodes and collate results at the end. 
Note that the script will continue to run until all jobs have completed. It may therefore 
be advisable to use [nohup](http://en.wikipedia.org/wiki/Nohup).


# Plotting the p-value distribution
